File: text/tests/aligno_test.py
#!/usr/bin/env python3

import os
from io import StringIO

import pytest

import aligno as subject
from aligno import aligno as subject_main

# Test data
SPACE_INDENTED = """
def hello():
    print("Hello, world!")
    if True:
        print("Nested block")
"""

TAB_INDENTED = """
def hello():
\tprint("Hello, world!")
\tif True:
\t\tprint("Nested block")
"""

MIXED_INDENTED = """
def hello():
    print("Hello, world!")
\tif True:
\t\tprint("Nested block")
"""

UNINDENTED = "def hello():\nprint('Hello, world!')\n"

COMMON_INDENTED = """
    def hello():
        print("Hello, world!")
        if True:
            print("Nested block")
"""

# Test detect_indent function
@pytest.mark.parametrize("input_text, expected", [
    (SPACE_INDENTED, (4, "s", 0)),
    (TAB_INDENTED, (1, "t", 0)),
    (MIXED_INDENTED, (1, "t", 0)),
    (UNINDENTED, (0, "", 0)),
    (COMMON_INDENTED, (4, "s", 1)),
])
def test_detect_indent(input_text, expected):
    assert subject.detect_indent(input_text) == expected

# Test apply_indent function
@pytest.mark.parametrize("input_text, indent_size, indent_type, min_level, expected", [
    (SPACE_INDENTED, 1, "t", 0, TAB_INDENTED),
    (TAB_INDENTED, 4, "s", 0, SPACE_INDENTED),
    (UNINDENTED, 2, "s", 1, "  def hello():\n  print('Hello, world!')\n"),
    (COMMON_INDENTED, 2, "s", 0, SPACE_INDENTED.replace("    ", "  ")),
])
def test_apply_indent(input_text, indent_size, indent_type, min_level, expected):
    assert subject.apply_indent(input_text, indent_size, indent_type, min_level).strip() == expected.strip()

# Test main function
def test_main_detect():
    input_stream = StringIO(SPACE_INDENTED)
    output_stream = StringIO()
    subject_main(istream=input_stream, ostream=output_stream, detect=True)
    assert output_stream.getvalue().strip() == "4s"

def test_main_apply():
    input_stream = StringIO(SPACE_INDENTED)
    output_stream = StringIO()
    subject_main("t", istream=input_stream, ostream=output_stream)
    assert output_stream.getvalue().strip() == TAB_INDENTED.strip()

@pytest.mark.parametrize("input_text", [
    SPACE_INDENTED,
    TAB_INDENTED,
    UNINDENTED,
    COMMON_INDENTED,
])
def test_main_default(input_text):
    input_stream = StringIO(input_text)
    output_stream = StringIO()
    expected = subject.apply_indent(input_text, *subject.parse_indent_code(subject.DEFAULT_INDENT)).strip()
    subject_main(istream=input_stream, ostream=output_stream, apply=True)
    assert output_stream.getvalue().strip() == expected

def test_detect_indent_list():
    input_lines = [
        "def hello():",
        "    print('Hello, world!')",
        "    if True:",
        "        print('Nested block')"
    ]
    assert subject.detect_indent(input_lines) == (4, "s", 0)

# Test edge cases
def test_empty_input():
    assert subject.detect_indent("") == (0, "", 0)
    assert subject.apply_indent("", 4, "s", 2) == "\n"

def test_single_line():
    assert subject.detect_indent("print('Hello')") == (0, "", 0)
    assert subject.apply_indent("print('Hello')", 2, "s", 1) == "  print('Hello')\n"

def test_mixed_indentation():
    mixed = "def foo():\n    print('bar')\n\tprint('baz')"
    assert subject.detect_indent(mixed) == (4, "s", 0)
    assert subject.apply_indent(mixed, 1, "t", 0) == "def foo():\n\tprint('bar')\nprint('baz')\n"

def test_large_indentation():
    large_indent = "def foo():\n        print('bar')"
    assert subject.detect_indent(large_indent) == (8, "s", 0)
    assert subject.apply_indent(large_indent, 2, "s", 1) == "  def foo():\n    print('bar')\n"

# Test invalid inputs
def test_invalid_indent_type():
    with pytest.raises(ValueError):
        subject.parse_indent_code("x4")

def test_negative_min_level():
    with pytest.raises(ValueError):
        subject.parse_indent_code("4s-1")

# Test parse_indent_code and format_indent_code
@pytest.mark.parametrize("indent_code, expected", [
    ("t", (1, "t", 0)),
    ("4s", (4, "s", 0)),
    ("2s1", (2, "s", 1)),
    ("t2", (1, "t", 2)),
])
def test_parse_indent_code(indent_code, expected):
    assert subject.parse_indent_code(indent_code) == expected

@pytest.mark.parametrize("indent_size, indent_type, min_level, expected", [
    (1, "t", 0, "t"),
    (4, "s", 0, "4s"),
    (2, "s", 1, "2s1"),
    (1, "t", 2, "t2"),
])
def test_format_indent_code(indent_size, indent_type, min_level, expected):
    assert subject.format_indent_code(indent_size, indent_type, min_level) == expected

# Test command-line interface
def test_cli_detect():
    input_stream = StringIO(SPACE_INDENTED)
    output_stream = StringIO()
    subject_main(istream=input_stream, ostream=output_stream, detect=True)
    assert output_stream.getvalue().strip() == "4s"

def test_cli_apply():
    input_stream = StringIO(SPACE_INDENTED)
    output_stream = StringIO()
    subject_main("t", istream=input_stream, ostream=output_stream, apply=True)
    assert output_stream.getvalue().strip() == TAB_INDENTED.strip()

def test_cli_default():
    os.environ.pop("FILETYPE", None)
    os.environ.pop("INDENT", None)
    input_stream = StringIO(SPACE_INDENTED)
    output_stream = StringIO()
    subject_main(istream=input_stream, ostream=output_stream, apply=True)
    assert output_stream.getvalue() == TAB_INDENTED

if __name__ == "__main__":
    pytest.main([__file__])

File: text/aligno_test.go
package main

import (
	"bytes"
	"io"
	"os"
	"strings"
	"testing"
)

// Use explicit string literals (with \n and no extra formatting from raw-string indents).
const (
	spaceIndented = "def hello():\n    print(\"Hello, world!\")\n    if True:\n        print(\"Nested block\")\n"
	tabIndented   = "def hello():\n\tprint(\"Hello, world!\")\n\tif True:\n\t\tprint(\"Nested block\")\n"
	mixedIndented = "def hello():\n    print(\"Hello, world!\")\n\tif True:\n\t\tprint(\"Nested block\")\n"
	unindented    = "def hello():\nprint('Hello, world!')\n"
	commonIndented = "    def hello():\n        print(\"Hello, world!\")\n        if True:\n            print(\"Nested block\")\n"
	bogusCommon = "\t    def hello():\n\t        print(\"Hello, world!\")\n\t        if True:\n\t            print(\"Nested block\")\n"
)

// Helper: create a temporary file containing the given content.
func stringToTempFile(t *testing.T, content string) *os.File {
	f, err := os.CreateTemp("", "stdin-temp")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := f.WriteString(content); err != nil {
		f.Close()
		t.Fatal(err)
	}
	if _, err := f.Seek(0, 0); err != nil {
		f.Close()
		t.Fatal(err)
	}
	return f
}

// Test detectIndent function
func TestDetectIndent(t *testing.T) {
	tests := []struct {
		input      string
		size       int
		indentType string
		minLevel   int
		shouldErr  bool
	}{
		// All nonempty lines must have a common indent.
		{spaceIndented, 4, "s", 0, false},
		{tabIndented, 1, "t", 0, false},
		{mixedIndented, 1, "t", 0, false},
		{unindented, 0, "", 0, false},
		{commonIndented, 4, "s", 1, false},
		{bogusCommon, 0, "", 0, true},
	}
	for _, test := range tests {
		size, typ, level, err := detectIndent(test.input)
		if test.shouldErr {
			if err == nil {
				t.Errorf("Expected error for input %q but got none", test.input)
			}
			continue
		}
		if err != nil {
			t.Errorf("Unexpected error for input %q: %v", test.input, err)
			continue
		}
		if size != test.size || typ != test.indentType || level != test.minLevel {
			t.Errorf("For input %q expected (%d, %q, %d) but got (%d, %q, %d)",
				test.input, test.size, test.indentType, test.minLevel,
				size, typ, level)
		}
	}
}

// Test applyIndent function
func TestApplyIndent(t *testing.T) {
	tests := []struct {
		input      string
		indentSize int
		indentType string
		minLevel   int
		expected   string
	}{
		// Converting from spaces to tabs.
		{spaceIndented, 1, "t", 0, tabIndented},
		// Converting from tabs to spaces.
		{tabIndented, 4, "s", 0, spaceIndented},
		// When the input has no indent.
		{unindented, 2, "s", 1, "  def hello():\n  print('Hello, world!')\n"},
		// With a common indent baseline.
		{commonIndented, 2, "s", 0, strings.ReplaceAll(spaceIndented, "    ", "  ")},
	}
	for _, test := range tests {
		output, err := applyIndent(test.input, test.indentSize, test.indentType, test.minLevel)
		if err != nil {
			t.Fatalf("Error applying indent: %v", err)
		}
		if output != test.expected {
			t.Errorf("Expected:\n%q\nGot:\n%q", test.expected, output)
		}
	}
}

// Test CLI detect functionality
func TestMainDetect(t *testing.T) {
	input := spaceIndented
	// When detected, aligno calls formatIndentCode and then if the result starts with "0", DEFAULT_INDENT is used.
	// For our input, detectIndent should yield (4, "s", 0) and then formatIndentCode returns "4s" (does not start with "0"),
	// so expected output is "4s\n"
	expected := "4s\n"

	oldStdin := os.Stdin
	oldStdout := os.Stdout
	defer func() {
		os.Stdin = oldStdin
		os.Stdout = oldStdout
	}()

	// Create temporary file for stdin.
	f := stringToTempFile(t, input)
	defer func() {
		f.Close()
		os.Remove(f.Name())
	}()
	os.Stdin = f

	// Use a pipe to capture stdout.
	r, w, err := os.Pipe()
	if err != nil {
		t.Fatal(err)
	}
	os.Stdout = w

	if err := aligno(true, false, ""); err != nil {
		t.Fatalf("Error while detecting indent: %v", err)
	}
	w.Close()

	var buf bytes.Buffer
	if _, err := io.Copy(&buf, r); err != nil {
		t.Fatal(err)
	}

	result := buf.String()
	if result != expected {
		t.Errorf("Expected %q but got %q", expected, result)
	}
}

// Test CLI apply functionality
func TestMainApply(t *testing.T) {
	input := spaceIndented
	expected := tabIndented

	oldStdin := os.Stdin
	oldStdout := os.Stdout
	defer func() {
		os.Stdin = oldStdin
		os.Stdout = oldStdout
	}()

	// Create a temporary file for stdin.
	f := stringToTempFile(t, input)
	defer func() {
		f.Close()
		os.Remove(f.Name())
	}()
	os.Stdin = f

	// Use a pipe for stdout.
	r, w, err := os.Pipe()
	if err != nil {
		t.Fatal(err)
	}
	os.Stdout = w

	err = aligno(false, true, "t")
	if err != nil {
		t.Fatalf("Error while applying indent: %v", err)
	}
	w.Close()

	var buf bytes.Buffer
	if _, err := io.Copy(&buf, r); err != nil {
		t.Fatal(err)
	}

	result := strings.TrimSuffix(buf.String(), "\n")
	if result != expected[:len(expected)-1] { // remove the trailing newline from expected for the comparison
		t.Errorf("Expected %q but got %q", expected, result+"\n")
	}
}

// Test empty input
func TestEmptyInput(t *testing.T) {
	// detectIndent on an empty string should return zeros.
	size, typ, level, err := detectIndent("")
	if err != nil {
		t.Errorf("Unexpected error on empty input: %v", err)
	}
	if size != 0 || typ != "" || level != 0 {
		t.Errorf("Expected (0,\"\",0) but got (%d,%q,%d)", size, typ, level)
	}

	// applyIndent on empty input should produce only the minimum indent plus a final newline.
	// For parameters (4, "s", 2), minIndent is 8 spaces.
	expected := "        \n"
	applyResult, err := applyIndent("", 4, "s", 2)
	if err != nil {
		t.Fatalf("Error during apply indent: %v", err)
	}

	if applyResult != expected {
		t.Errorf("Expected apply result %q but got %q", expected, applyResult)
	}
}

// Test single line
func TestSingleLine(t *testing.T) {
	line := "print('Hello')"
	// detectIndent on a single unindented line should return zeros.
	size, typ, level, err := detectIndent(line)
	if err != nil {
		t.Errorf("Unexpected error on single line: %v", err)
	}
	if size != 0 || typ != "" || level != 0 {
		t.Errorf("Expected (0,\"\",0) for single line but got (%d,%q,%d)", size, typ, level)
	}

	expected := "  print('Hello')\n"
	applyResult, err := applyIndent(line, 2, "s", 1)
	if err != nil {
		t.Fatalf("Error during apply indent: %v", err)
	}

	if applyResult != expected {
		t.Errorf("Expected apply result %q but got %q", expected, applyResult)
	}
}

// Test parseIndentCode and formatIndentCode
func TestParseFormatIndentCode(t *testing.T) {
	tests := []struct {
		indentCode string
		expected   struct {
			size       int
			indentType string
			minLevel   int
		}
	}{
		{"t", struct {
			size       int
			indentType string
			minLevel   int
		}{1, "t", 0}},
		{"4s", struct {
			size       int
			indentType string
			minLevel   int
		}{4, "s", 0}},
		{"2s1", struct {
			size       int
			indentType string
			minLevel   int
		}{2, "s", 1}},
		{"t2", struct {
			size       int
			indentType string
			minLevel   int
		}{1, "t", 2}},
	}

	for _, test := range tests {
		size, indentType, minLevel, err := parseIndentCode(test.indentCode)
		if err != nil {
			t.Fatalf("Error parsing indent code %q: %v", test.indentCode, err)
		}

		if size != test.expected.size || indentType != test.expected.indentType || minLevel != test.expected.minLevel {
			t.Errorf("Mismatch for %q. Expected (%d, %q, %d) but got (%d, %q, %d)",
				test.indentCode, test.expected.size, test.expected.indentType, test.expected.minLevel,
				size, indentType, minLevel)
		}

		formatted, err := formatIndentCode(size, indentType, minLevel)
		if err != nil {
			t.Fatalf("Error formatting indent code %q: %v", test.indentCode, err)
		}

		if formatted != test.indentCode {
			t.Errorf("Expected formatted result %q but got %q", test.indentCode, formatted)
		}
	}
}
